libsupport/include/katana/IndexID.h

```diff
@@ -51,26 +51,100 @@ struct IndexID : public OpaqueIDOrderedWithValue<_IDType, _Value> {
+    return *static_cast<_IDType*>(this);
+  }
+
+  _IDType& operator%=(const _IDType& rhs) { return *this %= rhs.value(); }
+  _IDType& operator%=(_Value val) {
+    this->value_ %= val;
+    return *static_cast<_IDType*>(this);
+  }
+
+  _IDType& operator>>=(const _IDType& rhs) { return *this >>= rhs.value(); }
+  _IDType& operator>>=(_Value val) {
+    this->value_ >>= val;
+    return *static_cast<_IDType*>(this);
+  }
+
+  _IDType& operator<<=(const _IDType& rhs) { return *this <<= rhs.value(); }
+  _IDType& operator<<=(_Value val) {
+    this->value_ <<= val;
+    return *static_cast<_IDType*>(this);
+  }
+
+  _IDType& operator&=(const _IDType& rhs) { return *this &= rhs.value(); }
+  _IDType& operator&=(_Value val) {
+    this->value_ &= val;
+    return *static_cast<_IDType*>(this);
+  }
 
+  _IDType& operator|=(const _IDType& rhs) { return *this |= rhs.value(); }
+  _IDType& operator|=(_Value val) {
+    this->value_ |= val;
+    return *static_cast<_IDType*>(this);
+  }
+
+  _IDType& operator^=(const _IDType& rhs) { return *this ^= rhs.value(); }
+  _IDType& operator^=(_Value val) {
+    this->value_ ^= val;
+    return *static_cast<_IDType*>(this);
+  }
+
+  _IDType operator+() const { return _IDType(+this->value_); }
+
+  _IDType operator-() const { return _IDType(-this->value_); }
+
+  _IDType operator~() const { return _IDType(~this->value_); }
+
+  _IDType operator+(const _IDType& rhs) const { return *this + rhs.value(); }
   _IDType operator+(_Value val) const { return _IDType(this->value() + val); }
 
   _IDType operator-(const _IDType& rhs) const { return *this - rhs.value(); }
-
   _IDType operator-(_Value val) const { return _IDType(this->value() - val); }
+
+  _IDType operator*(const _IDType& rhs) const { return *this * rhs.value(); }
+  _IDType operator*(_Value val) const { return _IDType(this->value() * val); }
+
+  _IDType operator/(const _IDType& rhs) const { return *this / rhs.value(); }
+  _IDType operator/(_Value val) const { return _IDType(this->value() / val); }
+
+  _IDType operator%(const _IDType& rhs) const { return *this % rhs.value(); }
+  _IDType operator%(_Value val) const { return _IDType(this->value() % val); }
+
+  _IDType operator>>(const _IDType& rhs) const { return *this >> rhs.value(); }
+  _IDType operator>>(_Value val) const { return _IDType(this->value() >> val); }
+
+  _IDType operator<<(const _IDType& rhs) const { return *this << rhs.value(); }
+  _IDType operator<<(_Value val) const { return _IDType(this->value() << val); }
+
+  _IDType operator&(const _IDType& rhs) const { return *this & rhs.value(); }
+  _IDType operator&(_Value val) const { return _IDType(this->value() & val); }
+
+  _IDType operator|(const _IDType& rhs) const { return *this | rhs.value(); }
+  _IDType operator|(_Value val) const { return _IDType(this->value() | val); }
+
+  _IDType operator^(const _IDType& rhs) const { return *this ^ rhs.value(); }
+  _IDType operator^(_Value val) const { return _IDType(this->value() ^ val); }
```

Comment:
I understand the need for the numeric operations, however I think the bit-level operations should be avoided. They are not operations on numbers. They are operations on bit vectors. C numeric types are both, but I don't see why we should follow that, it's a clear breach of encapsulation.

That said, I will not block the PR over this.

Reply:
I explicitly need `>>` and `<<`  because that's how the stl implements `{lower,upper}_bound`. I don't feel strongly about the other operations; I just added them because it seemed weird to define some bit operations and not others.

Reply:
STL has manual strength reduction!?!?!?! OMG. That's really bad. Oh well. Include them. It's C++, nothing is logical. ;-)

Reply:
> It's C++, nothing is logical. ;-)
But apparently everything is arithmetic!

Reply:
And yet they have no exponentiation operation. Sigh.